mutable_tree.go

```diff
@@ -226,22 +226,22 @@ func (tree *MutableTree) Load() (int64, error) {
 
 // Returns the version number of the latest version found
 func (tree *MutableTree) LoadVersion(targetVersion int64) (int64, error) {
-	roots, err := tree.ndb.getRoots()
-	if err != nil {
-		return 0, err
-	}
-	if len(roots) == 0 {
-		return 0, nil
-	}
 	latestVersion := int64(0)
 	var latestRoot []byte
-	for version, r := range roots {
-		tree.versions[version] = true
+	tree.ndb.traversePrefix([]byte(rootPrefix), func(k, v []byte) {
```

Comment:
We ought to be able to avoid traversing the entire prefix here. nodedb should expose an `traverseRange` which in turn can use `db.Iterator([]byte(rootPrefix), ndb.rootKey(targetVersion))` since we know the exact end point.

Come to think of it - do we really need even need to maintain a set of versions that exist? The only place this set is used is to 'fail early' when trying to get a previous tree version. `GetImmutable` will throw an error if the version is not found. I do not see why that isn't generally good enough. Then we could avoid this traversal completely (other than in loading the latest version).

Reply:
Comos-sdk use IAVL to implement its fundamental store. It implements prune strategy:
```
https://github.com/cosmos/cosmos-sdk/blob/2c6ed0fefc499410a7734ac93005f32a99b47b5b/store/iavlstore.go#L104
```
and
```
https://github.com/cosmos/cosmos-sdk/blob/2c6ed0fefc499410a7734ac93005f32a99b47b5b/store/iavlstore.go#L82
```
So the entire prefix here will not be too huge. But if some developers don't want to implement prune strategy here it is necessary to expose an traverseRange.
Thanks for your impressive comment.

Reply:
I currently don't prune (for reference I'm using it here: https://github.com/hyperledger/burrow/blob/develop/execution/state.go).

I'm not terribly worried about the size - it does load fairly quickly.

But my afterthought is probably now my main point:  I don't think we need the versions set at all, as in, I think we could remove this field: https://github.com/tendermint/iavl/blob/master/mutable_tree.go#L19 

Then all we need to do is load `targetVersion`. Later on rather than checking our versions set we can just handle this error gracefully: https://github.com/tendermint/iavl/blob/master/mutable_tree.go#L270. It means less loading up front - and I suspect most old versions never get accessed again.

For loading the latest version we'll need to traverse anyway.

Reply:
Would be happy to implement this slightly bigger change in a subsequent PR - what you have here is essential for my use case.

Reply:
I'm glad my pull request is helpful for your work. Thanks.